Sources/AWSSDKSwift/Services/S3/S3+multipart_API.swift

```diff
@@ -251,3 +212,55 @@ public extension S3 {
         }
     }
 }
+
+
+extension S3 {
+    /// used internally in multipartUpload, loads all the parts once the multipart upload has been initiated
+    func multipartUploadParts(_ input: CreateMultipartUploadRequest, uploadId: String, on eventLoop: EventLoop, inputStream: @escaping () -> EventLoopFuture<Data?>) -> Future<[S3.CompletedPart]> {
+        let promise = eventLoop.makePromise(of: [S3.CompletedPart].self)
+        var completedParts: [S3.CompletedPart] = []
+
+        // function uploading part of a file and queueing up upload of the next part
+        func multipartUploadPart(partNumber: Int, uploadId: String, body: Data? = nil) {
+            // create upload data future, if there is no data to load because this is the first time this is called create a succeeded future
+            let uploadResult : Future<[S3.CompletedPart]>
+            if let body = body {
+                let request = S3.UploadPartRequest(
+                    body: body,
+                    bucket: input.bucket,
+                    contentLength: Int64(body.count),
+                    key: input.key,
+                    partNumber: partNumber,
+                    requestPayer: input.requestPayer,
+                    sSECustomerAlgorithm: input.sSECustomerAlgorithm,
+                    sSECustomerKey: input.sSECustomerKey,
+                    sSECustomerKeyMD5: input.sSECustomerKeyMD5,
+                    uploadId: uploadId
+                )
+                // request upload future
+                uploadResult = self.uploadPart(request).hop(to: eventLoop).map { output -> [S3.CompletedPart] in
+                    let part = S3.CompletedPart(eTag: output.eTag, partNumber: partNumber)
+                    completedParts.append(part)
+                    return completedParts
+                }
+            } else {
+                uploadResult = eventLoop.makeSucceededFuture([])
+            }
+
+            // load data EventLoopFuture
+            let result = inputStream()
+                .and(uploadResult)
+                // upload data
+                .map { (data, parts) -> Void in
+                    guard let data = data, data.count > 0 else { return promise.succeed(parts)}
+                    multipartUploadPart(partNumber: partNumber+1, uploadId: uploadId, body: data)
+            }
+            result.whenFailure { error in
+                promise.fail(error)
+            }
+        }
+
```

Comment:
maybe we want to read once before going into recursion? If we don't get any data we could fail with a nothing to upload error. Only if we have a first part we could start the recursion:

Within the recursion we could upload(part) and inputStream(part + 1) concurrently. We already do this now but starting with an empty upload, which is IMHO hard to read later. Took me about 20min to get into this code.

Reply:
ok changed to send in first block of data